drake/core/Core.h

```diff
@@ -0,0 +1,106 @@
+#ifndef DRAKE_DRAKECORE_H
+#define DRAKE_DRAKECORE_H
+
+#include <string>
+#include <vector>
+#include <Eigen/Dense>
+
+namespace Drake {
+
+#include "Path.h"
+
+// useful refs on generic programming: http://www.generic-programming.org/languages/cpp/techniques.php
+//
+// Concept: Vector<ScalarType>.
+// Requirements:
+//  - Defines a VectorTraits<VectorType> with a static constant (or enum) RowsAtCompileTime (see Pendulum.h for an example).  Can be Eigen::DYNAMIC
+//  - implements size_t size(), which equals size_at_compile except when DYNAMIC
+//  - implements the copy constructor and the assignment operator= from Eigen::Matrix<ScalarType,size_at_compile,1> (e.g. using "copy and swap")
+//  - implements the typecast method operator Eigen::Matrix<ScalarType,size_at_compile,1>()
+//  - implements std::ostream& operator<<(std::ostream& os, const Vector& x)
+// (The methods/defs below make it possible (and hopefully easy) to simply use Eigen::Matrix types to model this concept)
+
+  template<typename T>
+  struct VectorTraits;
+
+  template<typename _Scalar, int _Rows, int _Cols, int _Options, int _MaxRows, int _MaxCols>
+  struct VectorTraits<Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols> > {
+    enum {
+      RowsAtCompileTime = _Rows
+    };
+  };
+
+  // In order to use Eigen types to model the Drake::Vector concept, they must accept only a single template parameter (ScalarType)
+  // The following structures provide a means for doing it. (e.g. via Drake::VectorBuilder<Rows>::VecType )
+
+  template <int Rows>
+  struct VectorBuilder {
+    template <typename ScalarType> using VecType = Eigen::Matrix<ScalarType,Rows,1>;
+  };
+  template <typename ScalarType> using EmptyVector = Eigen::Matrix<ScalarType,0,1>;
+  template <typename ScalarType> using UnusedVector = Eigen::Matrix<ScalarType,0,1>;
+
+
+  template <typename ScalarType, template <typename> class Vector1, template <typename> class Vector2>
+  class CombinedVector {
```

Comment:
I wonder if this should be variadically templated so that it can handle concatenations of arbitrarily many types.

In fact, isn't CombinedVector just a std::tuple with an assignment operator that takes an eigen matrix?